File: 3.py
from typing import List

class Student:
    def __init__(self, name: str, mtk: int, bahasa: int, ipa: int):
        self.name = name
        self.mtk = mtk
        self.bahasa = bahasa
        self.ipa = ipa
    
    def average(self) -> float:
        return (self.mtk + self.bahasa + self.ipa) / 3

class Classroom:
    def __init__(self, students: List[Student]):
        self.students = students
    
    def top_3(self) -> List[Student]:
        self.students.sort(key=lambda x: x.average(), reverse=True)
        return self.students[:3]
    
    def find_average(self, name: str) -> float:
        for student in self.students:
            if student.name == name:
                return student.average()
        return -1

def main():
    students = []
    name = ""
    mtk = 0
    bahasa = 0
    ipa = 0
    
    while True:
        name, mtk, bahasa, ipa = input().split()
        if name == "stop":
            break
        students.append(Student(name, int(mtk), int(bahasa), int(ipa)))
    
    classroom = Classroom(students)
    student_name = input()
    average = classroom.find_average(student_name)
    if average != -1:
        print("Nilai rata-rata dari", student_name, "adalah", average)
    top_3_students = classroom.top_3()
    for i, student in enumerate(top_3_students):
        print(student.name, "rangking",i+1, "dengan rata-rata", student.average())

if __name__ == "__main__":
    main()

File: 3/3.go
package main

import "fmt"

const NMAX = 10

type siswa struct {
	nama             string
	mtk, bahasa, ipa int
}

type kelas [NMAX]siswa

func main() {
	var data kelas
	var N, i int
	var nama string

	input_data(&data, &N)
	sort_data(&data, &N)
	fmt.Scanln(&nama)
	i = search_data(&data, &N, nama)
	if i != -1 {
		fmt.Println("Nilai rata-rata dari", data[i].nama, "adalah", rata_rata(&data[i]))

	}
	print_data(&data, &N)
	fmt.Println("Ok")
}

func input_data(T *kelas, N *int) {
	var nama string
	var mtk, bahasa, ipa int
	for i := 0; i < NMAX; i++ {
		fmt.Scanln(&nama, &mtk, &bahasa, &ipa)
		if nama == "stop" {
			break
		}
		(*T)[i].nama = nama
		(*T)[i].mtk = mtk
		(*T)[i].bahasa = bahasa
		(*T)[i].ipa = ipa
		*N = i + 1
	}
}

func rata_rata(M *siswa) float64 {
	var avg float64
	for i := 0; i < NMAX; i++ {
		avg = float64((*M).mtk+(*M).bahasa+(*M).ipa) / 3
	}
	return avg
}

func search_data(T *kelas, N *int, nama string) int {
	var i int
	for i = 0; i < *N; i++ {
		if (*T)[i].nama == nama {
			return i
		}
	}
	return -1
}

func sort_data(T *kelas, N *int) {
	for i := 0; i < *N; i++ {
		for j := 0; j < *N; j++ {
			if float64((*T)[i].mtk+(*T)[i].bahasa+(*T)[i].ipa) < float64((*T)[j].mtk+(*T)[j].bahasa+(*T)[j].ipa) {
				(*T)[j], (*T)[i] = (*T)[i], (*T)[j]
			}
		}
	}
}

func print_data(T *kelas, N *int) {
	var j int = 1
	for i := *N - 1; i > *N-4; i-- {
		fmt.Println((*T)[i].nama, "rangking", j, "dengan rata-rata", float64((*T)[i].mtk+(*T)[i].bahasa+(*T)[i].ipa)/3)
		j++
	}
}
